fix(hivemind): improve DHT initialization with robust fallback configurations

zh1p4ng opened the pull request:
fix： hydra.errors.InstantiationException: Error in call to target 'genrl_swarm.communication.hivemind.hivemind_backend.HivemindBackend': FileNotFoundError(2, 'No such file or directory')

fix(hivemind): improve DHT initialization with robust fallback configurations Fixes initialization failures with FileNotFoundError and P2P.create timeout errors. The new approach tries configurations in order of safety:
1. Local loopback (127.0.0.1) - most compatible
2. System auto-selection - let OS choose
3. Local loopback with await_ready disabled
4. All interfaces (0.0.0.0) TCP only
5. Original TCP+UDP configuration as last resort

zh1p4ng commented:
Error executing job with overrides: []
Traceback (most recent call last):
  File "/Users/zh1p4ng/Desktop/gen/rl-swarm/.venv/lib/python3.13/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 92, in _call_target
    return _target_(*args, **kwargs)
  File "/Users/zh1p4ng/Desktop/gen/rl-swarm/genrl-swarm/src/genrl_swarm/communication/hivemind/hivemind_backend.py", line 84, in __init__
    self.dht = DHT(
               ~~~^
        start=True,
        ^^^^^^^^^^^
    ...<2 lines>...
        **kwargs,
        ^^^^^^^^^
    )
    ^
  File "/Users/zh1p4ng/Desktop/gen/rl-swarm/.venv/lib/python3.13/site-packages/hivemind/dht/dht.py", line 87, in __init__
    self.run_in_background(await_ready=await_ready)
    ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zh1p4ng/Desktop/gen/rl-swarm/.venv/lib/python3.13/site-packages/hivemind/dht/dht.py", line 148, in run_in_background
    self.wait_until_ready(timeout)
    ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^
  File "/Users/zh1p4ng/Desktop/gen/rl-swarm/.venv/lib/python3.13/site-packages/hivemind/dht/dht.py", line 151, in wait_until_ready
    self._ready.result(timeout=timeout)
    ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "/Users/zh1p4ng/Desktop/gen/rl-swarm/.venv/lib/python3.13/site-packages/hivemind/utils/mpfuture.py", line 254, in result
    return super().result(timeout)
           ~~~~~~~~~~~~~~^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.13/3.13.3/Frameworks/Python.framework/Versions/3.13/lib/python3.13/concurrent/futures/_base.py", line 456, in result
    return self.__get_result()
           ~~~~~~~~~~~~~~~~~^^
  File "/opt/homebrew/Cellar/python@3.13/3.13.3/Frameworks/Python.framework/Versions/3.13/lib/python3.13/concurrent/futures/_base.py", line 401, in __get_result
    raise self._exception
FileNotFoundError: [Errno 2] No such file or directory

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/zh1p4ng/Desktop/gen/rl-swarm/genrl-swarm/src/genrl_swarm/runner/swarm_launcher.py", line 24, in <module>
    main()
    ~~~~^^
  File "/Users/zh1p4ng/Desktop/gen/rl-swarm/.venv/lib/python3.13/site-packages/hydra/main.py", line 94, in decorated_main
    _run_hydra(
    ~~~~~~~~~~^
        args=args,
        ^^^^^^^^^^
    ...<3 lines>...
        config_name=config_name,
        ^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/Users/zh1p4ng/Desktop/gen/rl-swarm/.venv/lib/python3.13/site-packages/hydra/_internal/utils.py", line 394, in _run_hydra
    _run_app(
    ~~~~~~~~^
        run=args.run,
        ^^^^^^^^^^^^^
    ...<5 lines>...
        overrides=overrides,
        ^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/Users/zh1p4ng/Desktop/gen/rl-swarm/.venv/lib/python3.13/site-packages/hydra/_internal/utils.py", line 457, in _run_app
    run_and_report(
    ~~~~~~~~~~~~~~^
        lambda: hydra.run(
        ^^^^^^^^^^^^^^^^^^
    ...<3 lines>...
        )
        ^
    )
    ^
  File "/Users/zh1p4ng/Desktop/gen/rl-swarm/.venv/lib/python3.13/site-packages/hydra/_internal/utils.py", line 223, in run_and_report
    raise ex
  File "/Users/zh1p4ng/Desktop/gen/rl-swarm/.venv/lib/python3.13/site-packages/hydra/_internal/utils.py", line 220, in run_and_report
    return func()
  File "/Users/zh1p4ng/Desktop/gen/rl-swarm/.venv/lib/python3.13/site-packages/hydra/_internal/utils.py", line 458, in <lambda>
    lambda: hydra.run(
            ~~~~~~~~~^
        config_name=config_name,
        ^^^^^^^^^^^^^^^^^^^^^^^^
        task_function=task_function,
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        overrides=overrides,
        ^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/Users/zh1p4ng/Desktop/gen/rl-swarm/.venv/lib/python3.13/site-packages/hydra/_internal/hydra.py", line 132, in run
    _ = ret.return_value
        ^^^^^^^^^^^^^^^^
  File "/Users/zh1p4ng/Desktop/gen/rl-swarm/.venv/lib/python3.13/site-packages/hydra/core/utils.py", line 260, in return_value
    raise self._return_value
  File "/Users/zh1p4ng/Desktop/gen/rl-swarm/.venv/lib/python3.13/site-packages/hydra/core/utils.py", line 186, in run_job
    ret.return_value = task_function(task_cfg)
                       ~~~~~~~~~~~~~^^^^^^^^^^
  File "/Users/zh1p4ng/Desktop/gen/rl-swarm/genrl-swarm/src/genrl_swarm/runner/swarm_launcher.py", line 17, in main
    game_manager = instantiate(cfg.game_manager)
  File "/Users/zh1p4ng/Desktop/gen/rl-swarm/.venv/lib/python3.13/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 226, in instantiate
    return instantiate_node(
        config, *args, recursive=_recursive_, convert=_convert_, partial=_partial_
    )
  File "/Users/zh1p4ng/Desktop/gen/rl-swarm/.venv/lib/python3.13/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 342, in instantiate_node
    value = instantiate_node(
        value, convert=convert, recursive=recursive
    )
  File "/Users/zh1p4ng/Desktop/gen/rl-swarm/.venv/lib/python3.13/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 347, in instantiate_node
    return _call_target(_target_, partial, args, kwargs, full_key)
  File "/Users/zh1p4ng/Desktop/gen/rl-swarm/.venv/lib/python3.13/site-packages/hydra/_internal/instantiate/_instantiate2.py", line 97, in _call_target
    raise InstantiationException(msg) from e
hydra.errors.InstantiationException: Error in call to target 'genrl_swarm.communication.hivemind.hivemind_backend.HivemindBackend':
FileNotFoundError(2, 'No such file or directory')
full_key: game_manager.communication
wandb: 
wandb: You can sync this run to the cloud by running:
wandb: wandb sync ./logs/wandb/offline-run-20250628_155739-ipy6vtez
>> An error was detected while running rl-swarm. See /Users/zh1p4ng/Desktop/gen/rl-swarm/logs for full logs.

0xnightwind commented on src/genrl_swarm/communication/hivemind/hivemind_backend.py:
```diff
@@ -123,6 +198,6 @@ def all_gather_object(self, obj: Any) -> Dict[str| int, Any]:
-        return str(self.dht.peer_id)
-    
+    def get_id(self) -> str:
+        """Get node ID"""
```
i feel like it's quite self explanatory so might not need this docstring

zh1p4ng replied:
Thanks for the feedback! 

- Added newline at EOF
-  Removed the docstring since the method name and type annotation already make it self-explanatory

The updated method is now:
def get_id(self) -> str:
    return str(self.dht.peer_id)

0xnightwind commented on src/genrl_swarm/communication/hivemind/hivemind_backend.py:
```diff
@@ -69,27 +69,102 @@ def __init__(
+                        **final_config,
+                    )
+                    
+                    time.sleep(2)  # Wait for bootstrap node to be ready
```
so is this duration derived from somewhere or is arbitary?

0xnightwind commented on src/genrl_swarm/communication/hivemind/hivemind_backend.py:
```diff
@@ -69,27 +69,102 @@ def __init__(
+                
+                # Clean up failed DHT instance
+                if self.dht:
+                    try:
+                        self.dht.shutdown()
```
im not sure if just the shutdown is fine, i hope this also does a full cleanup?

zh1p4ng commented:
Code Review Updates
﻿
I've addressed all the feedback points:
﻿
Completed Changes:
1. EOF & Docstring: Removed unnecessary docstring from `get_id()` and ensured EOF newline
2. Wait Time: Replaced arbitrary `time.sleep(2)` with configurable polling mechanism
3. Resource Cleanup: Implemented comprehensive DHT cleanup with graceful/force strategies
﻿
Additional Improvements:
- Added class constants for better maintainability
- Separated initialization logic into focused methods
- Improved error handling and type annotations
- Enhanced performance with better data structures
﻿
Testing:
- [x] Verified resource cleanup with multiple retry scenarios
- [x] Confirmed configurable wait times work as expected
﻿
Ready for re-review!

zh1p4ng commented:
After updating the rl-swarm repository, this issue has not been resolved, so it is still necessary to update this code
